features/42-cloud-function-log-isolation/api.md

```diff
@@ -0,0 +1,194 @@
+function types are invocable: hook, timer and lambda. More design is
+needed to support auth provider because it is consist of multiple functions.
+Handler can be invoked directly by calling the handler.
+
+On the server side, the list of all cloud function is stored in a struct
+called `registrationInfo`. The `registrationInfo` is per-plugin. But
+the result from `cloudfunc:list` should aggregate info from all plugins.
+
+Request:
+
+```json
+{
+  "action": "cloudfunc:list"
+}
+```
+
+Response:
+
+```json
+{
+  "result": { /* same as registrationInfo */
+    "hook": [
+      { /* same as pluginHookInfo */
+        "name": "hook_name",
+        "type": "note",
+        "trigger": "before_save"
+        "async": true
+      }
+    ],
+    "handler": [
+      { /* same as pluginHandlerInfo */
+        "auth_required": true,
+        "name": "handler_name",
+        "methods": ["POST", "GET"],
+        "key_required": true,
+        "user_required": true
+      }
+    ] 
+    "op": [
+      { /* same as map[string]interface{} in registrationInfo.Lambdas */
+        /* TODO */
+      }
+    ] 
+    "timer": [
+      {
+        "name": "",
+        "spec": ""
+      }
+    ] 
+    "provider": [
+      {
+        "type": "",
+        "name": ""
+      }
+    ] 
+  }
+}
+```
+
+### Invoke cloud function
+
+When calling this action, Skygear Server creates a plugin request struct and
+pass it to the plugin transport. This action returns to the client with data
+received from the plugin transport. 
+
+Request:
+
+```json
+{
+  "action": "cloudfunc:invoke",
+  "func": {
+    "kind": "hook",
+    "name": "hook_name",
+    "type": "note",  /* only for db hook */
+    "trigger": "before_save"  /* only for db hook */
+  }
+  "param": {
+    /* the param is specific to the cloud function */
+  }
+}
+```
+
+Response:
+
+```json
+{
+  "result": {
+    "response": {
+      /* the response is specific to the cloud function */
+    }
+  }
+}
+```
+
+## Client SDK Changes
+
+The client SDK will not be updated because the purpose of this proposal is for
+testing cloud function. Client SDK serves app development.
+
+## Skycli Changes
+
+Find what cloud functions can be invoked:
+
+```
+$ skycli cloud function list
+TYPE      NAME           RECORD TYPE    TRIGGER
+hook      add_info       note           before_save
+lambda    get_messages
+timer     collect_money
+```
+
+Invoking a cloud function without payload will open editing session with
+template or payload from previous invocation.
+
+```
+$ skycli cloud function invoke hook add_info --record-type note --trigger before_save
+```
+
+`--record-type` and `--trigger` is not required if the name is unique.
+
+An editing session is opened with some content to help the user invoke
+cloud function without entering boilerplate. The editing session should
+be prepopulated with either:
+
+- Templated params according to the record type and trigger
+- Payload from previous invocation for the same cloud function if any
```

Comment:
The saved previous payload is specific to that cloud function, so it will not be affected by other `skycli` commands nor other cloud function of the same type.

That is,

```
skycli cloud function invoke hook add_info --record-type note --trigger before_save
```

will be different from 

```
skycli cloud function invoke hook add_info --record-type note --trigger after_save
```